From 575718a46dcf08bbad72bacdc5989b26096b7c68 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:18:10 -0400
Subject: [fs] nfsd: clean up sparse endianness warnings in nfscache.c

Message-id: <1410898849-22126-7-git-send-email-bfields@redhat.com>
Patchwork-id: 95693
O-Subject: [Patch RHEL7 006/165] nfsd: clean up sparse endianness warnings in nfscache.c
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Jeff Layton <jlayton@primarydata.com>

We currently hash the XID to determine a hash bucket to use for the
reply cache entry, which is fed into hash_32 without byte-swapping it.
Add __force to make sparse happy, and add some comments to explain
why.

Signed-off-by: Jeff Layton <jlayton@primarydata.com>
Reviewed-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: b3d8d1284a8275f7e761df5fc5f80c464ecd23dd
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfscache.c b/fs/nfsd/nfscache.c
index e5e4675..bfb7857 100644
--- a/fs/nfsd/nfscache.c
+++ b/fs/nfsd/nfscache.c
@@ -218,7 +218,12 @@ static void
 hash_refile(struct svc_cacherep *rp)
 {
  hlist_del_init(&rp->c_hash);
- hlist_add_head(&rp->c_hash, cache_hash + hash_32(rp->c_xid, maskbits));
+ /*
+  * No point in byte swapping c_xid since we're just using it to pick
+  * a hash bucket.
+  */
+ hlist_add_head(&rp->c_hash, cache_hash +
+   hash_32((__force u32)rp->c_xid, maskbits));
 }
 
 /*
@@ -342,7 +347,11 @@ nfsd_cache_search(struct svc_rqst *rqstp, __wsum csum)
  struct hlist_head  *rh;
  unsigned int  entries = 0;
 
- rh = &cache_hash[hash_32(rqstp->rq_xid, maskbits)];
+ /*
+  * No point in byte swapping rq_xid since we're just using it to pick
+  * a hash bucket.
+  */
+ rh = &cache_hash[hash_32((__force u32)rqstp->rq_xid, maskbits)];
  hlist_for_each_entry(rp, rh, c_hash) {
   ++entries;
   if (nfsd_cache_match(rqstp, csum, rp)) {
-- 
1.7.1